src/audio/smart_amp/smart_amp_generic.c

```diff
@@ -9,194 +9,292 @@
 #include <sof/audio/format.h>
 #include <sof/audio/smart_amp/smart_amp.h>
 
-#ifdef CONFIG_GENERIC
+static void remap_s32_to_s32(struct smart_amp_mod_stream *src_mod, uint32_t frames,
+			     const struct audio_stream __sparse_cache *src,
+			     const int8_t *chan_map)
+{
+	int fr, ch;
+	int src_frag = 0;
+	int src_ch = audio_stream_get_channels(src);
+	int src_idx;
+	int32_t *src_ptr;
+	int32_t *mod_ptr = (int32_t *)src_mod->buf.data;
+
+	for (fr = 0 ; fr < frames ; fr++) {
+		for (ch = 0 ; ch < src_mod->channels; ch++) {
+			if (chan_map[ch] == -1) {
+				*mod_ptr = 0;
+			} else {
+				src_idx = src_frag + chan_map[ch];
+				src_ptr = audio_stream_read_frag_s32(src, src_idx);
+				*mod_ptr = *src_ptr;
+			}
+			mod_ptr++;
+		}
+		src_frag += src_ch;
+	}
```

Comment:
@singalsu @andrula-song is this using the optimal method now ?

Reply:
It would be better to write with frames without wrap helper to avoid audio_stream_read_frag_s32() usage for every sample read.

Reply:
the most waste part is if else, take 48 frame, 2 channel as example, if need 96 check for if/else, this is big cost.
better to use channel in first loop, and then check, chan_map[ch], if == -1, with one loop, else with another loop, like below:

	for (ch = 0 ; ch < src_mod->channels; ch++) {
		if (chan_map[ch] == -1) {
			for (fr = 0 ; fr < frames ; fr++) {
				*mod_ptr = 0;
				mod_ptr += src_mod->channels;
			}
		} else {
			for (fr = 0 ; fr < frames ; fr++) {
				src_idx = src_frag + chan_map[ch];
				src_ptr = audio_stream_read_frag_s16(src, src_idx);
				*mod_ptr = *src_ptr;
                                mod_ptr += src_mod->channels;
			}
			src_frag += src_ch;
		}
	}


this will have only 2 if/else, will save cycles a lot.
Note: this is just reference code, please have more investigation to get correct offset calculation.

Reply:
@btian1 I don't think that's equivalent to as the

Reply:
Thanks for the advice. I will do the follow-up optimization

Reply:
it is just for demo purpose for save if/else, as I said in the note, "need more investigation to get correct offset"

Reply:
just updated the commit for better buffer read/write strategy.